master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+package searcher
+
+import (
+	"math"
+	"sort"
+
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// AsyncHalvingSearch implements a search using the asynchronous successive halving algorithm
+// (ASHA). The experiment will run until the target number of trials have been completed
+// in the bottom rung and no further promotions can be made to higher rungs.
+type asyncHalvingSearch struct {
+	defaultSearchMethod
+	model.AsyncHalvingConfig
+
+	rungs      []*rung
+	trialRungs map[RequestID]int
+	// earlyExitTrials contains trials that exited early that are still considered in the search.
+	earlyExitTrials     map[RequestID]bool
+	totalStepsCompleted int
+	maxTrials           int
+	trialsCompleted     int
+}
+
+const ashaExitedMetricValue = math.MaxFloat64
+
+func newAsyncHalvingSearch(config model.AsyncHalvingConfig) SearchMethod {
+	rungs := make([]*rung, 0, config.NumRungs)
+	for id := 0; id < config.NumRungs; id++ {
+		compound := math.Pow(config.Divisor, float64(config.NumRungs-id-1))
+		stepsNeeded := max(int(float64(config.TargetTrialSteps)/compound), 1)
+		rungs = append(rungs,
+			&rung{
+				stepsNeeded:       stepsNeeded,
+				outstandingTrials: 0})
+	}
+
+	return &asyncHalvingSearch{
+		AsyncHalvingConfig: config,
+		rungs:              rungs,
+		trialRungs:         make(map[RequestID]int),
+		earlyExitTrials:    make(map[RequestID]bool),
+		maxTrials:          config.MaxTrials,
+	}
+}
+
+// promotions handles bookkeeping of validation metrics and returns a RequestID to promote if
+// appropriate.
+func (r *rung) promotionsAsync(requestID RequestID, metric float64, divisor float64) []RequestID {
+	// See if there is a trial to promote. We are increasing the total number of trials seen by 1; the
+	// number of best trials that definitely should have been promoted so far (numPromote) can only
+	// stay the same or increase by 1.
+	oldNumPromote := int(float64(len(r.metrics)) / divisor)
+	numPromote := int(float64(len(r.metrics)+1) / divisor)
+
+	// Insert the new trial result in the appropriate place in the sorted list.
+	insertIndex := sort.Search(
+		len(r.metrics),
+		func(i int) bool { return r.metrics[i].metric > metric },
+	)
+	promoteNow := insertIndex < numPromote
+
+	r.metrics = append(r.metrics, trialMetric{})
+	copy(r.metrics[insertIndex+1:], r.metrics[insertIndex:])
+	r.metrics[insertIndex] = trialMetric{
+		requestID: requestID,
+		metric:    metric,
+		promoted:  promoteNow,
+		closed:    false,
+	}
+
+	// If the new trial is good enough, it should be promoted immediately (whether or not numPromote
+	// changes). Otherwise, if numPromote changes, there is some other trial that should be promoted,
+	// unless it has been promoted already.
+	if promoteNow {
+		return []RequestID{requestID}
+	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
+		if !t.promoted {
+			t.promoted = true
+			return []RequestID{t.requestID}
+		}
+	}
```

Comment:
non-blocking:  I think we can simplify it a bit:
```
switch {
    case promoteNow:
        return []RequestId{requestID}
    case numPromote != oldNumPromote && !r.metrics[oldNumPromote].promoted:
        t := &r.metrics[oldNumPromote]
        t.promoted = true
        return []RequestID{t.requestID}
    default:
        return nil
}
```

Reply:
agree this is cleaner!